test/e2e/endpoints.go

```diff
@@ -0,0 +1,196 @@
+	done := make(chan bool)
+	running := true
+	go func() {
+		ok := true
+		for running {
+			endpoints, err := c.Endpoints(ns).Get(serviceName)
+			if err == nil {
+				if len(endpoints.Endpoints) != len(expectedEndpoints) {
+					glog.Warningf("Unexpected endpoints: %#v, expected %v", endpoints, expectedEndpoints)
+				} else if validateIPs(c, ns, expectedPort, expectedEndpoints, endpoints) {
+					break
+				}
+			} else {
+				glog.Warningf("Failed to get endpoints: %v", err)
+			}
+			time.Sleep(time.Second)
+		}
+		done <- ok
+	}()
+
+	select {
+	case result := <-done:
+		return result
+	case <-time.After(60 * time.Second):
+		glog.Errorf("Timed out waiting for endpoints.")
+		running = false
+		return false
+	}
+}
+
+func addPod(c *client.Client, ns, name string, labels map[string]string) error {
+	pod := &api.Pod{
+		ObjectMeta: api.ObjectMeta{
+			Name:   name,
+			Labels: labels,
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				{
+					Name:  "test",
+					Image: "kubernetes/pause",
+					Ports: []api.Port{{ContainerPort: 80}},
+				},
+			},
+		},
+	}
+	_, err := c.Pods(ns).Create(pod)
+	return err
+}
+
+func TestEndpoints(c *client.Client) bool {
+	serviceName := "endpoint-test"
+	ns := api.NamespaceDefault
+	labels := map[string]string{
+		"foo": "bar",
+		"baz": "blah",
+	}
+
+	defer func() {
+		err := c.Services(ns).Delete(serviceName)
+		if err != nil {
+			glog.Errorf("Failed to delete service: %v (%v)", serviceName, err)
+		}
+	}()
+
+	service := &api.Service{
+		ObjectMeta: api.ObjectMeta{
+			Name: serviceName,
+		},
+		Spec: api.ServiceSpec{
+			Port:          80,
+			Selector:      labels,
+			ContainerPort: util.NewIntOrStringFromInt(80),
+		},
+	}
+	if _, err := c.Services(ns).Create(service); err != nil {
+		glog.Errorf("Failed to create endpoint test service: %v", err)
+		return false
+	}
+	expectedPort := "80"
+
+	if !validateEndpoints(c, ns, serviceName, expectedPort, []string{}) {
+		return false
+	}
+
+	name1 := "test1"
+	if err := addPod(c, ns, name1, labels); err != nil {
+		glog.Errorf("Failed to create pod: %v", err)
+		return false
+	}
+	names := []string{name1}
+	defer func() {
+		for _, name := range names {
+			err := c.Pods(ns).Delete(name)
+			if err != nil {
+				glog.Errorf("Failed to delete pod: %v (%v)", name, err)
+			}
+		}
+	}()
+
+	if !validateEndpoints(c, ns, serviceName, expectedPort, names) {
+		return false
+	}
+
+	name2 := "test2"
+	if err := addPod(c, ns, name2, labels); err != nil {
+		glog.Errorf("Failed to create pod: %v", err)
+		return false
+	}
+	names = append(names, name2)
+
+	if !validateEndpoints(c, ns, serviceName, expectedPort, names) {
+		return false
+	}
+
+	if err := c.Pods(ns).Delete(name1); err != nil {
+		glog.Errorf("Failed to delete pod: %s", name1)
+		return false
+	}
+	names = []string{name2}
+
+	if !validateEndpoints(c, ns, serviceName, expectedPort, names) {
+		return false
+	}
+
+	return true
+}
+
+var _ = Describe("TestEndpoints", func() {
```

Comment:
I'm working on a services suite now.  Just got the DNS test passing and hope to have the ROService tests soon.  I'll submit that change soon so this can land in that suite.